### PR TITLE
Flask 2.x support

### PR DESCRIPTION
**Pull request by mmozos**
[__Removed__](https://github.com/cuducos/alchemydumps/pull/27#pullrequestreview-747151926).

### CODE REVIEW

**Comment by cuducos** [comment]
Hi. If I understand it correctly, you used Flask 2's changelog as the description of your PR. This is related, but does not help much. Can you edit it and clarify how your actual changes bring Flask 2 support?

**Comment by mmozos** [comment]
> Version 2.0.0
> 
> Released 2021-05-11
> 
>     * Drop support for Python 2 and 3.5.
> 
>     * Bump minimum versions of other Pallets projects: Werkzeug >= 2, Jinja2 >= 3, MarkupSafe >= 2, ItsDangerous >= 2, Click >= 8. Be sure to check the change logs for each project. For better compatibility with other applications (e.g. Celery) that still require Click 7, there is no hard dependency on Click 8 yet, but using Click 7 will trigger a DeprecationWarning and Flask 2.1 will depend on Click 8.
> 
>     * JSON support no longer uses simplejson. To use another JSON module, override app.json_encoder and json_decoder. #3555
> 
>     * The encoding option to JSON functions is deprecated. #3562
> 
>     * Passing script_info to app factory functions is deprecated. This was not portable outside the flask command. Use click.get_current_context().obj if it’s needed. #3552
> 
>     * The CLI shows better error messages when the app failed to load when looking up commands. #2741
> 
>     * Add sessions.SessionInterface.get_cookie_name() to allow setting the session cookie name dynamically. #3369
> 
>     * Add Config.from_file() to load config using arbitrary file loaders, such as toml.load or json.load. Config.from_json() is deprecated in favor of this. #3398
> 
>     * The flask run command will only defer errors on reload. Errors present during the initial call will cause the server to exit with the traceback immediately. #3431
> 
>     * send_file() raises a ValueError when passed an io object in text mode. Previously, it would respond with 200 OK and an empty file. #3358
> 
>     * When using ad-hoc certificates, check for the cryptography library instead of PyOpenSSL. #3492
> 
>     * When specifying a factory function with FLASK_APP, keyword argument can be passed. #3553
> 
>     * When loading a .env or .flaskenv file, the current working directory is no longer changed to the location of the file. #3560
> 
>     * When returning a (response, headers) tuple from a view, the headers replace rather than extend existing headers on the response. For example, this allows setting the Content-Type for jsonify(). Use response.headers.extend() if extending is desired. #3628
> 
>     * The Scaffold class provides a common API for the Flask and Blueprint classes. Blueprint information is stored in attributes just like Flask, rather than opaque lambda functions. This is intended to improve consistency and maintainability. #3215
> 
>     * Include samesite and secure options when removing the session cookie. #3726
> 
>     * Support passing a pathlib.Path to static_folder. #3579
> 
>     * send_file and send_from_directory are wrappers around the implementations in werkzeug.utils. #3828
> 
>     * Some send_file parameters have been renamed, the old names are deprecated. attachment_filename is renamed to download_name. cache_timeout is renamed to max_age. add_etags is renamed to etag. #3828, #3883
> 
>     * send_file passes download_name even if as_attachment=False by using Content-Disposition: inline. #3828
> 
>     * send_file sets conditional=True and max_age=None by default. Cache-Control is set to no-cache if max_age is not set, otherwise public. This tells browsers to validate conditional requests instead of using a timed cache. #3828
> 
>     * helpers.safe_join is deprecated. Use werkzeug.utils.safe_join instead. #3828
> 
>     * The request context does route matching before opening the session. This could allow a session interface to change behavior based on request.endpoint. #3776
> 
>     * Use Jinja’s implementation of the |tojson filter. #3881
> 
>     * Add route decorators for common HTTP methods. For example, @app.post("/login") is a shortcut for @app.route("/login", methods=["POST"]). #3907
> 
>     * Support async views, error handlers, before and after request, and teardown functions. #3412
> 
>     * Support nesting blueprints. #593, #1548, #3923
> 
>     * Set the default encoding to “UTF-8” when loading .env and .flaskenv files to allow to use non-ASCII characters. #3931
> 
>     * flask shell sets up tab and history completion like the default python shell if readline is installed. #3941
> 
>     * helpers.total_seconds() is deprecated. Use timedelta.total_seconds() instead. #3962
> 
>     * Add type hinting. #3973.



> Hi. If I understand it correctly, you used Flask 2's changelog as the description of your PR. This is related, but does not help much. Can you edit it and clarify how your actual changes bring Flask 2 support?

Edited `pyproject.toml` files with updated version constraints and regenerated ` poetry.lock` with the command: `poetry lock --no-update`

File: .gitignore [hunk 1]
```diff
@@ -1,6 +1,7 @@
 *.egg-info/
 *.pyc
 .coverage
+.idea/
```
Comment (cuducos) [inline .gitignore:4]: This looks like a user specific thing unrelated to the project. It best fits a [global `.gitignore` file](https://sebastiandedeyne.com/setting-up-a-global-gitignore-file/), not the project's `.gitignore` file.

```suggestion
```

File: pyproject.toml [hunk 2]
```diff
@@ -28,17 +28,17 @@ keywords = [
 exclude = ["tests/"]
 
 [tool.poetry.dependencies]
-python = "^3.6"
-Flask = "^1.1.2"
-Flask-SQLAlchemy = "^2.4.1"
-python-decouple = "^3.3"
-SQLAlchemy = "^1.3.16"
+python = ">=3.6.2,<4.0"
+Flask = ">=1.1.2"
```
Comment (cuducos) [inline pyproject.toml:32]: We should lock it to versions up to 2.x.x, so we don't accidentally install it in Flask 3 in the future (just as you did with Python).

File: pyproject.toml [hunk 3]
```diff
@@ -28,17 +28,17 @@ keywords = [
...
+Flask-SQLAlchemy = "^2.5.1"
+python-decouple = "^3.4"
+SQLAlchemy = "^1.4.23"
```
Comment (cuducos) [inline pyproject.toml:35]: It looks like this change breaks the Poetry dev environment. Maybe [related to this issue](https://github.com/python-poetry/poetry/issues/4402). But `poetry install` is not working here, at least (crashing with the error from the issue).

I would ask: why do we need to update SQLAlchemy and how it relates to offering support to Flask 2? Maybe this is a completely different issue to address.

**Review by cuducos — CHANGES_REQUESTED** [review]
First of all, I'm gonna delete the opening post description as it is irrelevant to the PR (it's just Flask 2 changelog, I guess; does not mention this project or the changes in the PR).

Second, you offered this solution, but we must add this test case (Flask 2) to our test suite to make sure it works now and it's gonna work in the future.

As you haven't said exactly what breaks when using this project with Flask 2, I'm assuming it does not break, that the only limitation is the version pinning. If that is the case, one strategy is to add Flask 1 and Flask 2 to the `tox.ini` file (for [example](https://github.com/cuducos/django-public-admin/blob/master/tox.ini)).

Third, I added minor comments inline.

Lastly, many thanks for the contribution, I"m just pushing you on the communication side to get context and understand what's really going on, ok?

**Comment by cuducos** [comment]
I just re-added the CI for tests in 889346fc875015cfa157be2eb1a1fe74e69d3f85. You can rebase with `main` or `git cherry-pick 889346fc875015cfa157be2eb1a1fe74e69d3f85` so we have the CI working for your PR : )

**Review by mmozos — COMMENTED** [review]
There is an error about running `black` under CI process

**Comment by cuducos** [comment]
> There is an error about running `black` under CI process

Can you share either a link to the error logs, or the error message?

**Comment by cuducos** [comment]
There's something else: tests tagged with Flask 1 are running with Flask 2 (e.g. [here](https://github.com/cuducos/alchemydumps/runs/3528548641?check_suite_focus=true#step:6:29))

**Comment by mmozos** [comment]
> > There is an error about running `black` under CI process
> 
> Can you share either a link to the error logs, or the error message?

Here it is: https://github.com/mmozos/alchemydumps/runs/3526643752?check_suite_focus=true#step:6:44

**Comment by cuducos** [comment]
> There is an error about running `black` under CI process

I see you update `black` and this might be a side-effect. How does updating Black relate to the objective of this PR, i.e., offering Flask 2.0 support?

**Comment by mmozos** [comment]
Done a rollback of Dev-dependencies and tagged Flask by adding manually before install, same error between 2.x branch (before last commits) and 1.x branch of Flask. What do you think about this test? throws when a backup restore task should be happen: https://github.com/mmozos/alchemydumps/runs/3536362796#step:6:58

File: .github/workflows/tests.yaml [hunk 4]
```diff
@@ -25,6 +26,7 @@ jobs:
     - name: Install dependencies
       run: |
         .venv/bin/pip install --upgrade pip
+        .venv/bin/poetry add "Flask==${{ matrix.flask-version }}"
```
Comment (cuducos) [inline .github/workflows/tests.yaml:29]: This will change the lock file, what is not a good practice I guess.

File: pyproject.toml [hunk 5]
```diff
@@ -28,8 +28,7 @@ keywords = [
 exclude = ["tests/"]
 
 [tool.poetry.dependencies]
-python = "^3.6"
-Flask = "^1.1.2"
```
Comment (cuducos) [inline pyproject.toml:32]: We cannot remove this line. Flask is a crucial dependency of this project.

File: tox.ini [hunk 6]
```diff
@@ -8,6 +8,7 @@ setenv =
     FLASK_APP = tests/integration/app.py
 deps =
     flask1: flask>=1.0.0,<2.0.0
+    flask2: flask>=2.0.0,<4.0.0
```
Comment (cuducos) [inline tox.ini:11]: Why 4.0.0?

```suggestion
    flask2: flask>=2.0.0,<3.0.0
```

**Review by cuducos — CHANGES_REQUESTED** [review]
I think the attempt to control Flask version with Poetry is not the ideal. Poetry is there to assure the minimum viable environment for the extension, but not to pin specific versions.

I still think the reference shared earlier is a better way. In that example, we do not use Poetry in the tests, instead, we [install dependencies manually in the `tox.ini`'s `deps` lines](https://github.com/cuducos/django-public-admin/blob/master/tox.ini#L17-L23).

This approach avoids changing the Poetry lock file for the tests, and makes the tests reproducible locally with a simple command call `$ tox`. What do you think?

**Comment by mmozos** [comment]
I am thinking about previous and actual error `tests/integration/test_commands.py:96: AssertionError`, to restore isn't a backup file mandatory thing?
```
FAILED tests/integration/test_commands.py::TestCommands::test_create_restore_remove - AssertionError: 0 != 2
```

**Comment by cuducos** [comment]
> to restore isn't a backup file mandatory thing?

yes, that's why [the test creates the backup before line 96](https://github.com/cuducos/alchemydumps/blob/5a03fa03e83531fffc3a5bf09322c4592e97f2e6/tests/integration/test_commands.py#L67-L70) ; )

**Comment by cuducos** [comment]
My advice from here is to make this PR as simple as possible. One way of doing that is removing all unnecessary changes, reverting these files to what they look like in the main branch:

- `README.md`
- `flask_alchemydumps/database.py`
- `setup.cfg`
- 

The only line edited in `pyproject.toml` should be something like:

```diff
- Flask = "^1.1.2"
+ Flask = ">=1.1.2"
```
Then regenerate `poetry.lock` from there — but it might continue being the same.

Finally, `tox.ini` should have most of the changes. I would keep the current diff, but install the local package (maybe with `pip install -e .`).

With these changes we eliminate a lot of side effects that might distract us from focusing on Flask 2.

The _Files changed_ tab should show a maximum of three files, I guess:
1. `pyproject.toml` where we allow Flask 1 or newer
2. `tox.ini` where we add tests to make sure nothing breaks in Flask 2
3. `poetry.lock` only if there are changes

--- 

If we get that to work locally, we can replicate these changes to the CI adding Flask 2 to `.github/workflows/tests.yaml`'s matrix:

```diff
-        flask-version:
-          - 1
+        flask-version:
+          - 1
+          - 2
```

And probably removing the `poetry install` (`tox`'s gonna take care of installing dependencies).

However, the changes to `.github/workflows/tests.yaml` can be a separate PR too. The important now is to run `$ tox` locally and have all tests passing in Flask 2 and Flask 1 in all Python supported versions : )

**Pull request closed without merging**